Source/Core/DiscIO/VolumeWii.cpp

```diff
@@ -521,4 +525,127 @@ bool VolumeWii::CheckBlockIntegrity(u64 block_index, const Partition& partition)
   return CheckBlockIntegrity(block_index, cluster, partition);
 }
 
+bool VolumeWii::EncryptGroup(u64 offset, u64 partition_data_offset,
+                             u64 partition_data_decrypted_size,
+                             const std::array<u8, AES_KEY_SIZE>& key, BlobReader* blob,
+                             std::array<u8, GROUP_TOTAL_SIZE>* out)
+{
+  std::vector<std::array<u8, BLOCK_DATA_SIZE>> unencrypted_data(BLOCKS_PER_GROUP);
+  std::vector<std::array<u8, BLOCK_HEADER_SIZE>> unencrypted_hashes(BLOCKS_PER_GROUP);
+
+  std::array<std::promise<bool>, BLOCKS_PER_GROUP> read_promises;
+  std::array<std::future<bool>, BLOCKS_PER_GROUP> h0_futures;
+  std::array<std::future<bool>, BLOCKS_PER_GROUP / 8> h1_futures;
+  std::future<bool> h2_future;
+
+  for (size_t i = 0; i < BLOCKS_PER_GROUP; ++i)
+  {
+    h0_futures[i] = std::async(std::launch::async,
+                               [&unencrypted_data, &unencrypted_hashes, &read_promises, i]() {
+                                 if (!read_promises[i].get_future().get())
+                                   return false;
+
+                                 for (u32 j = 0; j < 31; ++j)
+                                 {
+                                   mbedtls_sha1_ret(unencrypted_data[i].data() + j * 0x400, 0x400,
```

Comment:
There's a lot of magic going on from here on downwards; are there any useful structs to use with `offsetof` for this, or perhaps constants/`constexpr` values to help document where they come from/what they mean?

Reply:
I've added a struct (in a new commit, since the change also affects code that's in master).